File: terra/src/main/scala/com/dafttech/terra/game/world/items/persistence/GameObject.java
package com.dafttech.terra.game.world.items.persistence;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public abstract class GameObject {

    // ****FOLLOWING PERSISTENCE CODE**** May harm your brain

    public List<Field> annotatedFields = new LinkedList<Field>();

    public GameObject() {
        List<Field> fields = getAllDeclaredFields(this.getClass(), null);
        for (Field f : fields) {
            f.setAccessible(true);
            Persistent p = f.getAnnotation(Persistent.class);
            if (p == null) continue;

            annotatedFields.add(f);
        }
    }

    public int getPrototypeHash() {
        return getHashBase().hashCode();
    }

    public boolean isSamePrototype(Object obj) {
        if (obj instanceof Prototype) return getPrototypeHash() == ((Prototype) obj).hashCode();
        if (obj instanceof GameObject) return getPrototypeHash() == ((GameObject) obj).getPrototypeHash();

        return false;
    }

    public String getHashBase() {
        StringBuilder hashBuilder = new StringBuilder();

        hashBuilder.append(this.getClass().getCanonicalName());

        for (Field f : annotatedFields) {
            try {
                hashBuilder.append(f.get(this).toString());
            } catch (IllegalArgumentException | IllegalAccessException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
        }

        return hashBuilder.toString();
    }

    private static List<Field> getAllDeclaredFields(Class<?> targetClass, List<Field> fields) {
        if (fields == null) fields = new ArrayList<Field>();
        for (Field field : targetClass.getDeclaredFields())
            if (!fields.contains(field)) fields.add(field);
        if (targetClass.getSuperclass() != null) getAllDeclaredFields(targetClass.getSuperclass(), fields);
        return fields;
    }

    @Override
    protected Object clone() throws CloneNotSupportedException {
        // TODO Auto-generated method stub
        return super.clone();
    }

    public String getName() {
        return this.getClass().getCanonicalName();
    }

    public Prototype toPrototype() {
        Prototype proto = new Prototype();

        proto.className = this.getClass().getCanonicalName();

        for (Field f : annotatedFields) {
            try {
                /*
                 * if (!(f.get(this) instanceof Serializable))
                 * System.out.println("WARNING! Field " + f.getName() + " in " +
                 * this.getClass().getCanonicalName() +
                 * " is not Serializable and cannot be saved!");
                 */
                if (f.isAnnotationPresent(Persistent.class)) proto.values.put(f, f.get(this));
            } catch (IllegalArgumentException | IllegalAccessException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
        }

        return proto;
    }
}

File: terra/src/main/scala/com/dafttech/terra/game/world/items/persistence/GameObject.scala
package com.dafttech.terra.game.world.items.persistence

import java.lang.reflect.Field

object GameObject {
  private var currentUUID: Int = 0

  private def getAllDeclaredFields(targetClass: Class[_]): List[Field] =
    (targetClass.getDeclaredFields ++ Option(targetClass.getSuperclass).toList.flatMap(getAllDeclaredFields)).distinct.toList
}

// ****FOLLOWING PERSISTENCE CODE**** May harm your brain
abstract class GameObject() {
  val uuid: Int = { GameObject.currentUUID = GameObject.currentUUID + 1; GameObject.currentUUID }

  override def hashCode(): Int = uuid

  lazy val annotatedFields: List[Field] = {
    val fields: List[Field] = GameObject.getAllDeclaredFields(this.getClass).toList
    fields.flatMap { field =>
      field.setAccessible(true)
      val p = field.getAnnotation(classOf[Persistent])
      Option(p).map(_ => field).toList
    }
  }

  def getPrototypeHash: Int = getHashBase.hashCode

  def isSamePrototype(obj: Any): Boolean = obj match {
    case prototype: Prototype => getPrototypeHash == prototype.hashCode
    case gameObject: GameObject => getPrototypeHash == gameObject.getPrototypeHash
    case _ => false
  }

  def getHashBase: String = {
    val hashBuilder = new StringBuilder()

    hashBuilder.append(this.getClass.getCanonicalName)

    for (f <- annotatedFields) {
      try hashBuilder.append(f.get(this).toString)
      catch {
        case e@(_: IllegalArgumentException | _: IllegalAccessException) =>
          // TODO Auto-generated catch block
          e.printStackTrace()
      }
    }
    hashBuilder.toString
  }

  @throws[CloneNotSupportedException]
  override protected def clone(): AnyRef = super.clone()

  def getName: String = this.getClass.getCanonicalName

  def toPrototype: Prototype = {
    val proto = new Prototype
    proto.className = this.getClass.getCanonicalName

    for (f <- annotatedFields) {
      try {
        /*
               * if (!(f.get(this) instanceof Serializable))
               * System.out.println("WARNING! Field " + f.getName() + " in " +
               * this.getClass().getCanonicalName() +
               * " is not Serializable and cannot be saved!");
               */
        if (f.isAnnotationPresent(classOf[Persistent])) proto.addValue(f, f.get(this))
      } catch {
        case e@(_: IllegalArgumentException | _: IllegalAccessException) =>
          e.printStackTrace()
      }
    }

    proto
  }
}
